pyanaconda/bootloader/zipl.py

```diff
@@ -46,10 +46,12 @@ class ZIPL(BootLoader):
 
     @property
     def stage2_format_types(self):
-        if productName.startswith("Red Hat "):          # pylint: disable=no-member
-            return ["xfs", "ext4", "ext3", "ext2"]
-        else:
-            return ["ext4", "ext3", "ext2", "xfs"]
+        # If the product file_system_type is valid, use it by default
+        valid_fs = ["ext4", "ext3", "ext2", "xfs"]
+        if conf.storage.file_system_type in valid_fs:
+            valid_fs.insert(0, conf.storage.file_system_type)
```

Comment:
I don't like the repetitive code. There are two places where this property is used and only one of them is affected by this change, so it makes more sense to move this code to the `default_boot_fstype` property.

Reply:
My worry with having a `default_boot_fstype` is if someone sets that to a value that isn't supported by the bootloader Anaconda would do as instructed, but the system wouldn't be able to boot.

This logic permits a fall back to a boot fs that is compatible.

Reply:
The class `InstallerStorage` has a property `default_boot_fstype` that uses the first item from `stage2_format_types` if no default boot fstype is set.

```
@property
def default_boot_fstype(self):
    """The default filesystem type for the boot partition."""
    if self._default_boot_fstype:
        return self._default_boot_fstype

    return self.bootloader.stage2_format_types[0]
```

This property can be extended like this:

```
@property
def default_boot_fstype(self):
    """The default filesystem type for the boot partition."""
    if self._default_boot_fstype:
        return self._default_boot_fstype

    if self._default_fstype in self.bootloader.stage2_format_types:
        return self._default_fstype 

    return self.bootloader.stage2_format_types[0]
```

We set the default fstype to `conf.storage.file_system_type` in the function `create_storage` when we create the storage object, so the logic should be the same at the end.

This solution is more flexible and it looks like we might be able to remove the method `set_default_boot_fstype`. That would fix the issue with the unsupported default boot fstype that you mentioned, because it looks like this problem could happen with the `autopart` command.

Reply:
That is a much better solution!

I'll be out of the office for the next week or so... it may be a bit for me to reimplement.  Did you want to do it since you actually found the elegant path?

Reply:
Implemented at: https://github.com/rhinstaller/anaconda/pull/2026